src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs

```diff
@@ -169,38 +185,83 @@ public bool TryGetValue(TKey key, out TValue value)
             Debug.Assert(hashTableLocal.Length > 0);
+            TValue sentinel;
             int mask = hashTableLocal.Length - 1;
             int hashCode = GetKeyHashCode(key);
             int tableIndex = HashInt1(hashCode) & mask;
 
-            if (hashTableLocal[tableIndex] == null)
+            TValue examineEntry = hashTableLocal[tableIndex];
+            if ((examineEntry == null) || (examineEntry == _entryInProcessOfWritingSentinel))
             {
+                sentinel = Volatile.Read(ref _entryInProcessOfWritingSentinel);
+                if (sentinel != null && CompareKeyToValue(key, sentinel))
+                {
+                    value = sentinel;
+                    return true;
+                }
                 value = null;
                 return false;
             }
 
-            if (CompareKeyToValue(key, hashTableLocal[tableIndex]))
+            if (CompareKeyToValue(key, examineEntry))
             {
-                value = hashTableLocal[tableIndex];
+                value = examineEntry;
                 return true;
             }
 
             int hash2 = HashInt2(hashCode);
             tableIndex = (tableIndex + hash2) & mask;
-
-            while (hashTableLocal[tableIndex] != null)
+            examineEntry = hashTableLocal[tableIndex];
+            while ((examineEntry != null) && (examineEntry != _entryInProcessOfWritingSentinel))
             {
-                if (CompareKeyToValue(key, hashTableLocal[tableIndex]))
+                if (CompareKeyToValue(key, examineEntry))
                 {
-                    value = hashTableLocal[tableIndex];
+                    value = examineEntry;
                     return true;
                 }
                 tableIndex = (tableIndex + hash2) & mask;
+                examineEntry = hashTableLocal[tableIndex];
+            }
+
+            sentinel = Volatile.Read(ref _entryInProcessOfWritingSentinel);
+            if (sentinel != null && CompareKeyToValue(key, sentinel))
+            {
+                value = sentinel;
+                return true;
             }
+
             value = null;
             return false;
         }
 
+        /// <summary>
+        /// Spin and wait for a sentinel to disappear.
+        /// </summary>
+        /// <param name="hashtable"></param>
+        /// <param name="tableIndex"></param>
+        /// <returns>The value that replaced the sentinel, or null</returns>
+        TValue WaitForSentinelInHashtableToDisappear(TValue[] hashtable, int tableIndex)
+        {
+            TValue sentinel = Volatile.Read(ref _entryInProcessOfWritingSentinel);
+            if (sentinel == null)
+                return null;
+
+            TValue value = Volatile.Read(ref hashtable[tableIndex]);
+            while (true)
+            {
+                for (int i = 0; (i < 10000) && value == sentinel; i++)
+                {
+                    value = Volatile.Read(ref hashtable[tableIndex]);
```

Comment:
>Volatile.Read [](start = 28, length = 13)

This `Volatile.Read` is technically needed -otherwise it is valid to CSE it out of the loop and kill the loop, which would still have correct behavior, in a way, but not what you want.

Other `Volatile.Read`s seem unnecessary.